src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -204,6 +205,19 @@ private ExecutionContext CallContractInternal(ContractState contract, ContractMe
             if (NativeContract.Policy.IsBlocked(Snapshot, contract.Hash))
                 throw new InvalidOperationException($"The contract {contract.Hash} has been blocked.");
 
+            if (method.NoReentry)
+            {
+                if (_noReentryMethods.ContainsKey(contract.Hash))
+                {
+                    if (_noReentryMethods[contract.Hash].Contains(method.Name)) throw new InvalidOperationException($"Cannot Call Method {method} Of Contract {contract.Hash} in an reentry manner.");
+                }
+                else
+                {
+                    _noReentryMethods[contract.Hash] = new List<string>();
+                }
+                _noReentryMethods[contract.Hash].Add(method.Name);
```

Comment:
Using something like method.ID instead of method.Name would be a good choice.
Contracts may exsit different methods with same name.

Reply:
We use method.Name and parameter counter for now to identify a method now.